File: medium/337.house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
		std::pair<int, int> calc_money(TreeNode* current) {
    	if (current == nullptr) {
      	return {0, 0};
      }
      
      auto [left_taken, left_not_taken] = calc_money(current->left);
      auto [right_taken, right_not_taken] = calc_money(current->right);
    
			const int result_taken = current->val + left_not_taken + right_not_taken;
      const int result_not_taken =
      	std::max(left_taken, left_not_taken) 
        + std::max(right_taken, right_not_taken);
      
      return {result_taken, result_not_taken};
    }
     
    int rob(TreeNode* root) {
    		auto [taken, not_taken] = calc_money(root);
        return std::max(taken, not_taken);
    }
};

File: medium/337.house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return 0,0
            not_take_left, take_left = dfs(node.left)
            not_take_right, take_right = dfs(node.right)
            #take current node
            res_taken = node.val + not_take_left + not_take_right
            #not take current node, we have a choice
            res_not_taken = max(not_take_left, take_left) + max(not_take_right, take_right)
            return res_not_taken, res_taken
        not_taken, taken = dfs(root)
        return max(not_taken, taken)
                